### PR TITLE
fix: truncates to token details descriptions

### PR DESCRIPTION
**Pull request by lynnshaoyu**
Truncates token descriptions longer than 400 characters.
https://uniswaplabs.atlassian.net/browse/WEB-778?atlOrigin=eyJpIjoiNTBiOWExNDU1OThhNGRlNGJmMTg2OGFhNWY4OTQxYjQiLCJwIjoiaiJ9

Screen Shot 2022-08-23 at 2 28 59 PM
Screen Shot 2022-08-23 at 2 28 52 PM

### CODE REVIEW

File: src/components/Tokens/TokenDetails/TokenDetail.tsx [hunk 1]
```diff
@@ -174,6 +169,98 @@ const MissingData = styled.div`
...
+        <ResourcesContainer>
+          <Resource name={'Etherscan'} link={`https://etherscan.io/address/${address}`} />
+          <Resource name={'Protocol Info'} link={`https://info.uniswap.org/#/tokens/${address}`} />
+        </ResourcesContainer>
+      </AboutContainer>
+    )
+  } else {
+    const tokenDescriptionArray = tokenDetailData.description.split(' ')
+    const shouldTruncate = tokenDescriptionArray.length > 60
+    const tokenDescriptionTruncated = shouldTruncate ? tokenDescriptionArray.slice(0, 59).join(' ') + '...' : undefined
```
Comment (vm) [inline src/components/Tokens/TokenDetails/TokenDetail.tsx:227]: use the constant for the 59

File: src/components/Tokens/TokenDetails/TokenDetail.tsx [hunk 2]
```diff
@@ -174,6 +169,98 @@ const MissingData = styled.div`
...
+  &:focus {
+    color: ${({ theme }) => darken(0.1, theme.textSecondary)};
+    cursor: pointer;
+  }
+`
+
+type TokenDetailDataProps = {
+  description: string | null
+  homepageUrl: string | null
+  twitterName: string | null
```
Comment (vm) [inline src/components/Tokens/TokenDetails/TokenDetail.tsx:203]: string?

File: src/components/Tokens/TokenDetails/TokenDetail.tsx [hunk 3]
```diff
@@ -174,6 +169,98 @@ const MissingData = styled.div`
...
+  padding: 12px 0px;
+
+  &:hover,
+  &:focus {
+    color: ${({ theme }) => darken(0.1, theme.textSecondary)};
+    cursor: pointer;
+  }
+`
+
+type TokenDetailDataProps = {
```
Comment (vm) [inline src/components/Tokens/TokenDetails/TokenDetail.tsx:200]: TokenDetailData

File: src/components/Tokens/TokenDetails/TokenDetail.tsx [hunk 4]
```diff
@@ -174,6 +169,98 @@ const MissingData = styled.div`
...
+  text-overflow: ellipsis;
+  max-width: 100%;
+  max-height: fit-content;
+`
+
+const TruncateDescriptionButton = styled.div`
+  color: ${({ theme }) => theme.textSecondary};
+  font-weight: 400;
+  font-size: 14px;
+  flex: none;
```
Comment (vm) [inline src/components/Tokens/TokenDetails/TokenDetail.tsx:190]: why is this necessary?

File: src/components/Tokens/TokenDetails/TokenDetail.tsx [hunk 5]
```diff
@@ -174,6 +169,98 @@ const MissingData = styled.div`
...
+    const tokenDescriptionTruncated = shouldTruncate ? tokenDescriptionArray.slice(0, 59).join(' ') + '...' : undefined
+
+    return (
+      <AboutContainer>
+        <AboutHeader>
+          <Trans>About</Trans>
+        </AboutHeader>
+        {!shouldTruncate && tokenDetailData.description}
+        {shouldTruncate && isDescriptionTruncated && (
+          <TokenDescriptionContainer>
```
Comment (vm) [inline src/components/Tokens/TokenDetails/TokenDetail.tsx:236]: this and the next container can be consolidated with ternaries inline

File: src/components/Tokens/TokenDetails/TokenDetail.tsx (hunk 2 above)
Comment (vm) [reply src/components/Tokens/TokenDetails/TokenDetail.tsx:203]: actually can leave this, i didn't realize this was the data type not the props.

File: src/components/Tokens/TokenDetails/TokenDetail.tsx (hunk 3 above)
Comment (lynnshaoyu) [reply src/components/Tokens/TokenDetails/TokenDetail.tsx:200]: done

File: src/components/Tokens/TokenDetails/TokenDetail.tsx [hunk 6]
```diff
@@ -174,6 +169,98 @@ const MissingData = styled.div`
...
+        </NoInfoAvailable>
+        <ResourcesContainer>
+          <Resource name={'Etherscan'} link={`https://etherscan.io/address/${address}`} />
+          <Resource name={'Protocol Info'} link={`https://info.uniswap.org/#/tokens/${address}`} />
+        </ResourcesContainer>
+      </AboutContainer>
+    )
+  } else {
+    const tokenDescriptionArray = tokenDetailData.description.split(' ')
+    const shouldTruncate = tokenDescriptionArray.length > 60
```
[the remark this answers is not in the document]
Comment (lynnshaoyu) [reply src/components/Tokens/TokenDetails/TokenDetail.tsx:226]: done

File: src/components/Tokens/TokenDetails/TokenDetail.tsx (hunk 2 above)
Comment (lynnshaoyu) [reply src/components/Tokens/TokenDetails/TokenDetail.tsx:203]: sgtm

File: src/components/Tokens/TokenDetails/TokenDetail.tsx (hunk 4 above)
Comment (lynnshaoyu) [reply src/components/Tokens/TokenDetails/TokenDetail.tsx:190]: ur right its not deleted

File: src/components/Tokens/TokenDetails/TokenDetail.tsx (hunk 5 above)
Comment (lynnshaoyu) [reply src/components/Tokens/TokenDetails/TokenDetail.tsx:236]: sounds good, done

File: src/components/Tokens/TokenDetails/TokenDetail.tsx [hunk 7]
```diff
@@ -174,6 +169,103 @@ const MissingData = styled.div`
   gap: 20px;
 `
 
+export const AboutContainer = styled.div`
+  display: flex;
```
Comment (vm) [inline src/components/Tokens/TokenDetails/TokenDetail.tsx:173]: why do you need this flex column?

File: src/components/Tokens/TokenDetails/TokenDetail.tsx [hunk 8]
```diff
@@ -174,6 +169,103 @@ const MissingData = styled.div`
...
+        <ResourcesContainer>
+          <Resource name={'Etherscan'} link={`https://etherscan.io/address/${address}`} />
+          <Resource name={'Protocol Info'} link={`https://info.uniswap.org/#/tokens/${address}`} />
+        </ResourcesContainer>
+      </AboutContainer>
+    )
+  } else {
+    const tokenDescriptionArray = tokenDetailData.description.split(' ')
+    const shouldTruncate = tokenDescriptionArray.length > TRUNCATE_WORDCOUNT
+    const tokenDescriptionTruncated = shouldTruncate
```
Comment (vm) [inline src/components/Tokens/TokenDetails/TokenDetail.tsx:228]: `tokenDescription = shouldTruncate ? tokenDescriptionArray.slice(0, TRUNCATE_WORDCOUNT - 1).join(' ') + '...' : tokenDetailData.description

File: src/components/Tokens/TokenDetails/TokenDetail.tsx [hunk 9]
```diff
@@ -174,6 +169,103 @@ const MissingData = styled.div`
...
+      ? tokenDescriptionArray.slice(0, TRUNCATE_WORDCOUNT - 1).join(' ') + '...'
```
Comment (vm) [inline src/components/Tokens/TokenDetails/TokenDetail.tsx:229]: `${tokenDescriptionArray.slice(0, TRUNCATE_WORDCOUNT - 1).join(' ')}...`

File: src/components/Tokens/TokenDetails/TokenDetail.tsx (hunk 7 above)
Comment (lynnshaoyu) [reply src/components/Tokens/TokenDetails/TokenDetail.tsx:173]: ur right not needed removed

File: src/components/Tokens/TokenDetails/TokenDetail.tsx (hunk 9 above)
Comment (lynnshaoyu) [reply src/components/Tokens/TokenDetails/TokenDetail.tsx]: updated to character count due to fred instructions to include paragraph breaks

File: src/components/Tokens/TokenDetails/TokenDetail.tsx [hunk 10]
```diff
@@ -174,6 +169,106 @@ const MissingData = styled.div`
...
+type TokenDetailData = {
+  description: string | null
+  homepageUrl: string | null
+  twitterName: string | null
+}
+
+export function AboutSection({ address, tokenDetailData }: { address: string; tokenDetailData: TokenDetailData }) {
+  const [isDescriptionTruncated, setIsDescriptionTruncated] = useState(true)
+
+  if (!tokenDetailData || !tokenDetailData.description) {
```
Comment (vm) [inline src/components/Tokens/TokenDetails/TokenDetail.tsx:210]: DMd about consolidating the structure with below

Comment (lynnshaoyu) [reply src/components/Tokens/TokenDetails/TokenDetail.tsx]: done

File: src/components/Tokens/TokenDetails/TokenDetail.tsx [hunk 11]
```diff
@@ -174,6 +169,100 @@ const MissingData = styled.div`
...
+
+const TRUNCATE_CHARACTER_COUNT = 400
+
+type TokenDetailData = {
+  description: string | null
+  homepageUrl: string | null
+  twitterName: string | null
+}
+
+const truncateDescription = (desc: string) => {
```
Comment (vm) [inline src/components/Tokens/TokenDetails/TokenDetail.tsx:207]: if < 400 char, return string
else:
description.substr(0, description.lastIndexOf(""))

File: src/components/Tokens/TokenDetails/TokenDetail.tsx [hunk 12]
```diff
@@ -238,8 +238,8 @@ export function AboutSection({ address, tokenDetailData }: { address: string; to
           <Trans>No token information available</Trans>
         </NoInfoAvailable>
       )}
-      {!shouldTruncate && tokenDescription}
```
Comment (vm) [inline src/components/Tokens/TokenDetails/TokenDetail.tsx]: rm

File: src/components/Tokens/TokenDetails/TokenDetail.tsx [hunk 13]
```diff
@@ -238,8 +238,8 @@ export function AboutSection({ address, tokenDetailData }: { address: string; to
...
       <TokenDescriptionContainer>
+        {!shouldTruncate && tokenDescription}
```
Comment (vm) [inline src/components/Tokens/TokenDetails/TokenDetail.tsx:242]: rm

File: src/components/Tokens/TokenDetails/TokenDetail.tsx [hunk 14]
```diff
@@ -238,8 +238,8 @@ export function AboutSection({ address, tokenDetailData }: { address: string; to
...
         {shouldTruncate && (
           <div>
```
Comment (vm) [inline src/components/Tokens/TokenDetails/TokenDetail.tsx:244]: rm

**Comment by infredible** [comment]
1. Line height for `body` (the text in token description) should be 24px
2. Padding between paragraph and "About" should be 16px (currently 12px)

File: src/components/Tokens/TokenDetails/TokenDetail.tsx [hunk 15]
```diff
@@ -225,18 +309,7 @@ export default function LoadedTokenDetail({ address }: { address: string }) {
...
-            </AboutHeader>
-            <NoInfoAvailable>
-              <Trans>No token information available</Trans>
-            </NoInfoAvailable>
-            <ResourcesContainer>
-              <Resource name={'Etherscan'} link={`https://etherscan.io/address/${address}`} />
-              <Resource name={'Protocol Info'} link={`https://info.uniswap.org/#/tokens/${address}`} />
-            </ResourcesContainer>
-          </AboutSection>
+          <AboutSection address={address} tokenDetailData={tokenDetailData as TokenDetailData} />
```
Comment (vm) [inline src/components/Tokens/TokenDetails/TokenDetail.tsx:310]: why do you need to cast?

Comment (lynnshaoyu) [reply src/components/Tokens/TokenDetails/TokenDetail.tsx:310]: it has more props than we need

Comment (lynnshaoyu) [reply src/components/Tokens/TokenDetails/TokenDetail.tsx:310]: visual studio complains when we dont cast

Comment (vm) [reply src/components/Tokens/TokenDetails/TokenDetail.tsx:311]: you cna make a new object without the extra props

File: src/components/Tokens/TokenDetails/TokenDetail.tsx [hunk 16]
```diff
@@ -174,6 +169,94 @@ const MissingData = styled.div`
...
+      </AboutHeader>
+      {(!tokenDetailData || !tokenDetailData.description) && (
+        <NoInfoAvailable>
+          <Trans>No token information available</Trans>
+        </NoInfoAvailable>
+      )}
+      <TokenDescriptionContainer>
+        {tokenDescription}
+        <TruncateDescriptionButton onClick={() => setIsDescriptionTruncated(!isDescriptionTruncated)}>
+          {isDescriptionTruncated ? <Trans>Read more</Trans> : <Trans>Hide</Trans>}
```
Comment (vm) [inline src/components/Tokens/TokenDetails/TokenDetail.tsx:245]: @infredible should be say "show less" or "collapse" instead of "hide"? it only truncates, doesn't hide.

Comment (infredible) [reply src/components/Tokens/TokenDetails/TokenDetail.tsx:245]: I covered this w/ andy and he preferred the succintness. Appreciate the feedback here tho

**Comment by infredible** [comment]
We use sentence-casing and not title-casing
CleanShot 2022-08-23 at 12 10 57@2x

this should be "Protocol info" and "Contract address"

**Comment by lynnshaoyu** [comment]
> We use sentence-casing and not title-casing CleanShot 2022-08-23 at 12 10 57@2x
> 
> this should be "Protocol info" and "Contract address"

done

File: src/components/Tokens/TokenDetails/TokenDetail.tsx [hunk 17]
```diff
@@ -193,6 +276,11 @@ export default function LoadedTokenDetail({ address }: { address: string }) {
   const networkBadgebackgroundColor = chainInfo?.backgroundColor
   const filterNetwork = useAtomValue(filterNetworkAtom)
   const tokenDetailData = useTokenDetailQuery(address, chainIdToChainName(filterNetwork))
+  const relevantTokenDetailData = (({ description, homepageUrl, twitterName }) => ({
```
Comment (vm) [inline src/components/Tokens/TokenDetails/TokenDetail.tsx:279]: this is clever tbh i can't think of a better way to do this without an external library or multiple lines

Comment (vm) [reply src/components/Tokens/TokenDetails/TokenDetail.tsx:279]: name this `aboutTokenDetailData` since relevant is vague

**Review by vm — APPROVED** [review]

File: src/components/Tokens/TokenDetails/TokenDetail.tsx (hunk 17 above)
Comment (lynnshaoyu) [reply src/components/Tokens/TokenDetails/TokenDetail.tsx:279]: 🙏🥲💖
